sdks/python/apache_beam/io/gcp/bigtableio_test.py

```diff
@@ -0,0 +1,218 @@
+    bundles.sort()
+    print('len(bundles) = {}'.format(len(bundles)))
+
+    self.assertEqual(len(list(bundles)), len(KEYS_1))
+
+  def _key_bytes(self, key):
+    return bytes(key) if sys.version_info < (3, 0) else bytes(key, 'utf8')
+
+  @mock.patch.object(Table, 'read_rows')
+  def test_read(self, mock_read_rows):
+
+    pos_start = 672496
+    row_count = 400
+
+    def _mock_read_list():
+      for i in range(pos_start, pos_start + row_count):
+        yield PartialRowData(self._key_bytes('beam_key%07d' % i))
+
+    mock_read_rows.return_value = _mock_read_list()
+    bigtable = BigtableSource(self.project_id, self.instance_id, self.table_id)
+    rows = list(bigtable.read(bigtable.get_range_tracker()))
+    self.assertEqual(len(rows), row_count)
+    for row in rows:
+      self.assertIsInstance(row, PartialRowData)
+      self.assertNotEqual(row.row_key, b'')
+
+  @mock.patch.object(Table, 'sample_row_keys')
+  @mock.patch.object(Table, 'read_rows')
+  def test_read_small_table(self, mock_read_rows, mock_sample_row_keys):
+    row_count = 10000
+
+    def _mock_read_rows(): # 12.2 KB
+      for i in range(0, row_count):
+        yield PartialRowData(self._key_bytes('beam_key%07d' % i))
+
+    mock_sample_row_keys.return_value = list(self._mock_sample_keys())
+    mock_read_rows.return_value = _mock_read_rows()
+    source = BigtableSource(self.project_id, self.instance_id, self.table_id)
+
+    for split_bundle in source.split(None):
+      range_tracker = source.get_range_tracker(split_bundle.start_position, split_bundle.stop_position)
+      rows = list(source.read(range_tracker))
+      self.assertEqual(len(rows), row_count)
+
+      for row in rows:
+        self.assertIsInstance(row, PartialRowData)
+
+  def _mocking_read_rows(self, **kwargs):  # 12.2 KB
+    start_key = kwargs['start_key'] if kwargs['start_key'] is not None else b''
+    end_key = kwargs['end_key']
+
+    from apache_beam.io.range_trackers import LexicographicKeyRangeTracker
+
+    index_start = 0
+    index_stop = RANGES_DICT[b''][1]
+    if start_key != b'':
+      fraction = LexicographicKeyRangeTracker.position_to_fraction(start_key)
+      index_start = int(fraction * index_stop)
+    elif end_key is not None:
+      fraction = LexicographicKeyRangeTracker.position_to_fraction(end_key)
+      index_stop = int(fraction * index_stop)
+
+    for i in range(index_start, index_stop):
+      yield PartialRowData(self._key_bytes('{}\x00{:07d}'.format(start_key, i)))
+
+  @mock.patch.object(Table, 'sample_row_keys')
+  @mock.patch.object(Table, 'read_rows')
+  def test_read_table(self, mock_read_rows, mock_sample_row_keys):
+    mock_sample_row_keys.return_value = list(self._mock_sample_keys(KEYS_2))
+    mock_read_rows.side_effect = self._mocking_read_rows
+    source = BigtableSource(self.project_id, self.instance_id, self.table_id)
+
+    # TODO: Need to implement mock-reader and row counter
+
+    bundles = [b for b in source.split(None)]
```

Comment:
Please also add tests similar to test_progress, test_read_reentrant, test_read_after_splitting here: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/textio_test.py (and any other test related to progress or splitting that I missed).

Reply:
As BigtableSource is no longer used, this one does not apply.